### DISCUSSION TITLE
QZ with WebServer

### DISCUSSION THREAD

**Discussion by greghesp** [cagnulein/qdomyos-zwift#747] [General]
Has anyone been able to get this working?

I'm trying to get QZ running in Docker with the WebServer functionality, however at the moment I'm a little unclear if this has been included in the latest build?  It seems there is a option in the settings on Android for this, yet there are docs to get it running too: https://github.com/cagnulein/qdomyos-zwift/blob/master/docs/40_web_socket_api.md

I've tried building it in Docker, although struggling.  So far I've got down to line `RUN /opt/Qt/5.12.12/bin/qmake` which then fails with `ERROR: Unknown module(s) in QT: websockets` but currently unsure if i even need to do this now?

```dockerfile
FROM debian:latest
MAINTAINER cagnulein

ENV MAKEFLAGS -j8
WORKDIR /usr/local/src

# Install QT 5.12.2
RUN apt -y update
RUN apt -y upgrade
RUN apt -y install wget
RUN apt-get install xz-utils
RUN wget -c https://download.qt.io/official_releases/qt/5.12/5.12.12/single/qt-everywhere-src-5.12.12.tar.xz 
RUN ls -alh
RUN tar -xvf qt-everywhere-src-5.12.12.tar.xz 
RUN apt -y install build-essential libfontconfig1-dev libdbus-1-dev libfreetype6-dev libicu-dev libinput-dev libxkbcommon-dev libsqlite3-dev libssl-dev libpng-dev libjpeg-dev libglib2.0-dev

RUN apt -y install bluez libgbm-dev
RUN apt -y install libudev-dev libinput-dev libts-dev libxcb-xinerama0-dev libxcb-xinerama0 gdbserver libqt5websockets5-dev
RUN apt -y install libegl1-mesa libegl1-mesa-dev libgles2-mesa libgles2-mesa-dev
#RUN apt -y install wiringpi 
RUN apt -y install libnfc-bin libnfc-dev fonts-texgyre libts-dev
RUN apt -y install libbluetooth-dev bluez-tools gstreamer1.0-plugins* libgstreamer1.0-dev libgstreamer-plugins-base1.0-dev libopenal-data libopenal1 libopenal-dev pulseaudio
RUN apt -y install gstreamer1.0-qt5
RUN apt -y install libasound2-dev libavcodec-dev libavformat-dev libswscale-dev gstreamer1.0-tools libgstreamer-plugins-*
RUN apt -y install qtdeclarative5-dev
RUN apt -y install libvlc-dev


RUN mkdir build
RUN cd build
RUN ../qt-everywhere-src-5.12.12/configure -v -opengl es2 -eglfs -no-gtk -opensource -confirm-license -release -reduce-exports -force-pkg-config -nomake examples -no-compile-examples -skip qtwayland -skip qtwebengine -no-feature-geoservices_mapboxgl -qt-pcre -no-pch -ssl -evdev -system-freetype -fontconfig -glib -prefix /opt/Qt/5.12.12 -qpa eglfs
RUN make
RUN make install

# Install QT Httpserver
RUN apt-get -y install git
RUN cd ~
RUN git clone https://github.com/qt-labs/qthttpserver
RUN cd ~/qthttpserver/src/3rdparty/http-parser
RUN wget https://raw.githubusercontent.com/nodejs/http-parser/main/http_parser.h
RUN wget https://raw.githubusercontent.com/nodejs/http-parser/main/http_parser.c
RUN cd ~/qthttpserver/src
RUN /opt/Qt/5.12.12/bin/qmake # Please note if you compiled Qt you need to specify /opt/Qt/5.12.12/bin/qmake
RUN make
RUN make install

# Install QDomyos-Zwift
RUN cd ~
RUN git clone https://github.com/cagnulein/qdomyos-zwift.git
RUN cd ~/qdomyos-zwift
RUN git submodule update --init ~/qdomyos-zwift/src/smtpclient/
RUN cd ~/qdomyos-zwift/src
RUN /opt/Qt/5.12.12/bin/qmake # Please note if you compiled Qt you need to specify /opt/Qt/5.12.12/bin/qmake
# make -j4 # Remove -j4 if you compiled on Raspberry Pi Zero
```

**Comment by cagnulein** [comment]
the docker was 1 year old and needs to be fixed. the web server is in master you have just to synchronize the docker file adding the missing modules. i can do it but i need time or you can do it and do a pr after :)

**Reply by greghesp** [reply]
TBH I've totally rewritten the dockerfile from the repo.  Do you know what the missing modules are?  From what I can tell I've added QT and the QT modules in the dockerfile which installed fine, but building fails as it can't find them

**Reply by cagnulein** [reply]
i'm outside now but you can double check the modules from the github workflow here 
i will be more precise when i will be home

**Reply by greghesp** [reply]
No worries.  Think i'm getting somewhere, its currently building..

```dockerfile
FROM ubuntu:latest
MAINTAINER cagnulein

ARG DEBIAN_FRONTEND=noninteractive
ENV TZ=Europe/Moscow
ENV MAKEFLAGS -j8
WORKDIR /usr/local/src

RUN apt-get update && apt-get install -y tzdata
RUN apt -y update
RUN apt -y upgrade
RUN apt update -y && apt-get install -y git qt5-default libqt5bluetooth5 libqt5widgets5 libqt5positioning5 libqt5xml5 qtconnectivity5-dev qtpositioning5-dev libqt5charts5-dev libqt5charts5 qt5-default libqt5networkauth5-dev libqt5websockets5* libxcb-randr0-dev libxcb-xtest0-dev libxcb-xinerama0-dev libxcb-shape0-dev libxcb-xkb-dev build-essential
RUN cd ~
RUN git clone https://github.com/cagnulein/qdomyos-zwift.git
RUN ls -alh
RUN cd ~/qdomyos-zwift

RUN git submodule update --init ~/qdomyos-zwift/src/smtpclient/
RUN git submodule update --init src/qmdnsengine/
RUN cd src
RUN qmake
RUN make-j4  
```

Only issue I have now is figuring out how to pass bluetooth through to Docker :D

```
Wed Mar 30 18:48:32 2022 1648655312434 Debug: bluetooth.cpp void bluetooth::debug(const QString&) "no bluetooth dongle found!"
"no bluetooth dongle found!"
```

**Reply by cagnulein** [reply]
great! About the Bluetooth I guess you need to run it as privileged mode. I never did it but it should work
Let me know how it will go

**Reply by greghesp** [reply]
I'm trying to get it running as a Dockerfile but having some issues for some reason.  
Previously i spun up an Ubuntu container and just ran these manually which worked, no idea why its failing though here using Dockerfile.  Any ideas?

```
#13 [ 9/15] RUN ls -alh
#13 sha256:4b5dfb1969e228b79ed0f0d06393eddbc97c0197582b068ad75b1e5672597867
#13 0.361 total 16K
#13 0.361 drwxr-xr-x  1 root root 4.0K Mar 30 20:31 .
#13 0.361 drwxr-xr-x  1 root root 4.0K Mar 16 20:07 ..
#13 0.361 drwxr-xr-x 17 root root 4.0K Mar 30 20:31 qdomyos-zwift
#13 DONE 0.4s

#14 [10/15] RUN cd ~/qdomyos-zwift
#14 sha256:c875dc5e19a73f6e654af2feb63f14be8a945e8eefe4882f4cd64962dd90d399
#14 0.437 /bin/sh: 1: cd: can't cd to /root/qdomyos-zwift
#14 ERROR: executor failed running [/bin/sh -c cd ~/qdomyos-zwift]: exit code: 2
```

**Reply by cagnulein** [reply]
maybe you were logged with a different account? now you're root on that container. Just a thought

**Reply by greghesp** [reply]
Ah using `RUN cd ....` does not work as you'd expect it to with Dockerfile. Needs to be `WORKDIR ....`

**Reply by cagnulein** [reply]
so it's working now @greghesp ?

**Reply by greghesp** [reply]
The Dockerfile appears to be complete and working now (full file below)
I just need to figure out how to pass Bluetooth through to it now.  I'm on a Windows client where I can't figure out how to do it, and I want to deploy into Docker on Home Assistant running on a NUC, which I also can't figure out how to do it yet :D

```dockerfile
FROM ubuntu:latest
MAINTAINER cagnulein

ARG DEBIAN_FRONTEND=noninteractive
ENV TZ=Europe/Moscow
ENV MAKEFLAGS -j8
WORKDIR /usr/local/src

RUN apt-get update && apt-get install -y tzdata
RUN apt -y update
RUN apt -y upgrade
RUN apt update -y && apt-get install -y git qt5-default libqt5bluetooth5 libqt5widgets5 libqt5positioning5 libqt5xml5 qtconnectivity5-dev qtpositioning5-dev libqt5charts5-dev libqt5charts5 qt5-default libqt5networkauth5-dev libqt5websockets5* libxcb-randr0-dev libxcb-xtest0-dev libxcb-xinerama0-dev libxcb-shape0-dev libxcb-xkb-dev build-essential

RUN git clone https://github.com/cagnulein/qdomyos-zwift.git
WORKDIR /usr/local/src/qdomyos-zwift

RUN git submodule update --init src/smtpclient/
RUN git submodule update --init src/qmdnsengine/
WORKDIR /usr/local/src/qdomyos-zwift/src
RUN qmake
RUN make -j4  
```

**Reply by cagnulein** [reply]
committed https://github.com/cagnulein/qdomyos-zwift/commit/ce7ad4aa8330bc8978bdda875e8d5748e321d3e6 , thanks

**Reply by greghesp** [reply]
Nice.  I might still need to make a few more tweaks, but I should be able to commit

**Reply by greghesp** [reply]
Looks like I wont be able to test the bluetooth integration with my Dockerfile until tomorrow.

Can't seem to find a way to pass through the device in Windows, and on MacOS, Docker is in a virtualised environment so doesn't have access to the MacOS Bluetooth integration.

Have bought a bluetooth dongle for my NUC which is running Home Assistant, so hopefully I can run it in there

**Comment by greghesp** [comment]
@cagnulein I ended up spinning an instance up on a Pi just to play around with first.

I can see its connected to the bike fine:
![image](https://user-images.githubusercontent.com/768908/161296973-0917cad4-5209-4d34-bc60-5164e55ed8ec.png)

However I'm unable to connect to the WS Server thats supposed to be running on it?

`Error: connect ECONNREFUSED 192.168.1.166:6666`

```
Now installed you need to compile like say in PR https://github.com/cagnulein/qdomyos-zwift/pull/252 and issue https://github.com/cagnulein/qdomyos-zwift/issues/572 template/debug in the same directory of source file of QDomyos-Zwift.

cp -r ~/qdomyos-zwift/src/templates/debug ~/qdomyos-zwift/src/.
cp -r ~/qdomyos-zwift/src/templates/debug/* ~/qdomyos-zwift/src/.
```

Done this

```
template_inner_QZWS_enabled=true
template_inner_QZWS_folders=:/inner_templates//chartjs
template_inner_QZWS_ips=192.168.1.166
template_inner_QZWS_port=6666
template_inner_QZWS_type=WebServer
```

File above has been edited too

**Reply by cagnulein** [reply]
did you build the qt_httpserver for arm too?

**Reply by greghesp** [reply]
I did not. I thought it was all included in master without the need to build separately, or did I misunderstand?

> the web server is in master you have just to synchronize the docker file adding the missing modules

**Reply by cagnulein** [reply]
the library is external: check this for reference https://github.com/cagnulein/qdomyos-zwift/issues/572#issuecomment-1011053445

**Reply by cagnulein** [reply]
in the master there is all the code that use the library. the QT's pro file, if it doesn't see the library, it just disabled the webserver on QZ

**Reply by greghesp** [reply]
Ah OK.  Does QT still need building as well as qt_httpserver?

(Sorry, come from a JS background so all a little different :D)

**Reply by cagnulein** [reply]
nope, not on linux. i patched the qt on ios and android because i found some bugs on the bluetooth part. but works ok out of the box on linux

**Comment by ttonyh** [comment]
@greghesp Have you been able to get the Docker file to work now? What about the webserver/websocket functionality? Based on your Dockerfile example (https://github.com/cagnulein/qdomyos-zwift/discussions/747#discussioncomment-2473018), with some modifications, I was able to fully compile and run the qdomyos-zwift binary and it detected my bike; but I could not get it to connect.

**Reply by cagnulein** [reply]
@ttonyh Share a debug log in a ticket and I will help you

**Reply by ttonyh** [reply]
Thank you Roberto; I have several questions; I will open up a new discussion.
